lib/Dialect/FIRRTL/Transforms/ResolvePaths.cpp

```diff
@@ -0,0 +1,221 @@
+//===- ResolvePaths.cpp - Resolve path operations ---------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file contains the ResolvePathsPass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotationHelper.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+struct PathResolver {
+  PathResolver(CircuitOp circuit, InstanceGraph &instanceGraph)
+      : circuit(circuit), symbolTable(circuit), instanceGraph(instanceGraph),
+        hierPathCache(circuit, symbolTable),
+        builder(OpBuilder::atBlockBegin(circuit->getBlock())) {}
+
+  /// This function will find the operation targeted and create a hierarchical
+  /// path operation if needed. If the target is resolved, the op will either
+  /// be a reference to the HierPathOp, or null if no HierPathOp was needed.
+  LogicalResult resolveHierPath(Location loc, const AnnoPathValue &target,
+                                FlatSymbolRefAttr &op) {
+
+    // We want to root this path at the top level module, or in the case of an
+    // unreachable module, we can settle for as high as we can get.
+    auto module = target.ref.getModule();
+    if (!target.instances.empty())
+      module = target.instances.front()->getParentOfType<FModuleLike>();
```

Comment:
Should this terminate at public (/DUT) modules?

Just running to top for now sounds good (and things can truncate the paths as needed?).

Reply:
I don't think the way these paths work makes sense with more than one public module, since the HierPathOps are created with a certain top-level module in mind.  I think the other way of representing path operations (where it can only reference local hardware) would make a lot more sense here.